package-res/resources/web/pentaho/visual/base/View.js

```diff
@@ -92,39 +89,105 @@ define([
+     * listen to the {@link pentaho.visual.base.events.DidCreate|"did:create"} event.
      *
      * @return {Promise} A promise that is fulfilled when the visualization
      * is completely rendered. If the visualization is in an invalid state, the promise
      * is immediately rejected.
+     *
+     * @fires "will:update"
+     * @fires "did:update"
+     * @fires "rejected:update"
+     * @fires "did:create"
      */
-    render: function() {
-      var me = this;
+    update: function() {
+      var me = this,
+          model = this.model,
+          isFirstRender = !this._element;
+
       return new Promise(function(resolve, reject) {
-        try {
-          var validationErrors = me._validate();
-          if(!validationErrors) {
-            Promise.resolve(me._render()).then(resolve, reject);
-          } else {
-            reject(validationErrors);
-          }
-        } catch(e) {
-          reject(e.message);
+        var willUpdate;
+        if(model._hasListeners(WillUpdate.type)) {
+          willUpdate = new WillUpdate(model);
+          model._emitSafe(willUpdate);
         }
+
+        var result = willUpdate && willUpdate.isCanceled
+          ? ActionResult.reject(willUpdate.cancelReason)
+          : me._doUpdate();
+
+        if(result.error) {
+          if(model._hasListeners(RejectedUpdate.type))
+            model._emitSafe(new RejectedUpdate(model, result.error));
+
+          reject(result.error);
+        } else {
+          if(model._hasListeners(DidUpdate.type))
+            model._emitSafe(new DidUpdate(model));
+
+          if(isFirstRender && model._hasListeners(DidCreate.type))
+            model._emitSafe(new DidCreate(model, me._element));
+
+          resolve();
+        }
+
       });
     },
 
     /**
+     * Updates a visualization.
+     *
+     * If the visualization is valid, the visualization element will be created on the first update
+     * and proceed with the visualization update; otherwise, it will be rejected and prevent the update.
+     *
+     * @return {pentaho.lang.ActionResult} The result object.
+     * @protected
+     */
+    _doUpdate: function() {
+      var result;
+      try {
+        var validationErrors = this._validate();
+        if(!validationErrors) {
+          if(!this._element) this.prepare();
+
+          this._update();
+          result = ActionResult.fulfill();
+
+        } else {
+          var error = "View update was rejected:\n - " +
+            (Array.isArray(validationErrors) ? validationErrors.join("\n - ") : validationErrors);
+          result = ActionResult.reject(error);
+        }
+
+      } catch(e) {
+        result = ActionResult.reject(e.message);
+      }
+
+      return result;
+    },
+
+    /**
      * Called before the visualization is discarded.
      */
     dispose: function() {
       this._element = null;
     },
 
     /**
+     * Called before the first valid visualization update.
+     */
+    prepare: function() {
```

Comment:
This should not be a public method, right?

Anyway, we should use the same naming as the one chosen for the DidCreate event and call this method something like `_willCreate` ?

Also, I'm not sure we should require the dom node to be created before _update is called, cause that would kind of defeat all of the effort we had for allowing the viz code to build the node — in `_update`, it may be fetching the code that builds the dom node...

So, it might be enough that a call to `_update` that does not result in a `_domNode` being set rejects the update (and does not emit the `did:create` event).

We may enforce a bit more control by creating a `_setDomNode(.)` method?

So I'd remove this method.